### ISSUE TITLE
Scaling for loss layers

### ISSUE THREAD

**Issue by piiswrong** [dmlc/mxnet#487]
Our current solution is to leave the loss as-is, and let the user decide the scale using grad_rescale. The Mathematical standard is to divide it by batch_size*num_output so that loss and learning_rate don't change when you change batch_size or num_output. 

Consider scaling the loss by batch_size*num_output by default?

**Comment by pluskid** [comment]
For `batch_size`, the gradient is already divided by `batch_size` in the updating rule of optimizers.

**Comment by piiswrong** [comment]
This is only for FeedForward model. I'm thinking about pushing it into the default setting of optimizers or loss_ops to make it work for every model.

**Comment by pluskid** [comment]
Yes I agree. In the Julia side I exposed more information to an optimizer, including `batch_size`, so rescaling is automatically done for all optimizers (see [here](https://github.com/dmlc/MXNet.jl/blob/master/src/optimizer.jl#L214)). And an optimizer is model-agnostic so should work with any models besides FeedForward. I could back-port that to the Python side if necessary.

For output size, are you talking about single loss operation with multi-dimension output (like vector valued regression or image segmentation), or multiple loss operations (like those deep models where a softmax loss is connected also to lower layer hidden representations)? It seems to me that the situation here is much more complicated to handle *automatically* for all general cases. Also, though the user might be tweaking the batch size a lot, changing the output dimension might not be a so common thing. I think it is good to have automatically scaling for specific losses like multi-output softmax-loss, or square loss for vector outputs, etc. But still leave the parameter grad_rescale to allow users to do customization.

**Comment by piiswrong** [comment]
I agree. Please do a patch if you have time. Otherwise I can do it later.
We can standardize scaling for mxnet provided multi-dimension output loss layers and recommend people do the same for custom layers.
This can help avoid dangerously large gradients when you have a hundred output and use a common learning rate of 0.01.

**Comment by mli** [comment]
please aware that if using the BSP model for distributed training, the batch size is multiplied by num_workerss, see https://github.com/dmlc/mxnet/blob/master/python/mxnet/model.py#L744

doc https://github.com/dmlc/mxnet/blob/master/doc/developer-guide/multi_node.md#kvstore

**Comment by pluskid** [comment]
Is it viable to unify the handling of mini batch for the distributed training and single-node multi-device training? For example, to let different devices process a whole mini-batch as in distributed training.

**Comment by mli** [comment]
the current `batch_size` means the number of examples a process will read each time. it somewhat controls two important hyper-parameters:

- `num_ex_per_device` the number of examples feed into a single device each time. it's limited by the device memory capacity
- `num_ex_per_update` the number of examples used for each weight updating. this is a trade-off between system performance (often in the distributed setting) and convergence speed.

we'd better make the relationship clearer

**Comment by tqchen** [comment]
closing for inactitivy
